src/core/options-watcher.cpp

```diff
@@ -0,0 +1,152 @@
+    }
+    catch( ... )
+    {
+        LOG_DEBUG( "Failures taking lock or stoping thread while destructing options_watcher" );
+    }
+}
+
+void options_watcher::register_option( rs2_option id, std::shared_ptr< option > option )
+{
+    registered_option opt = { option, 0.0f }; //Option actual value will be queried in start if needed.
+
+    {
+        std::lock_guard< std::mutex > lock( _mutex );
+        _options[id] = std::move( opt );
+    }
+
+    if( should_start() )
+        start();
+}
+
+void options_watcher::unregister_option( rs2_option id )
+{
+    {
+        std::lock_guard< std::mutex > lock( _mutex );
+        _options.erase( id );
+    }
+
+    if( should_stop() )
+        stop();
+}
+
+rsutils::subscription options_watcher::subscribe( callback && cb )
+{
+    rsutils::subscription ret = _on_values_changed.subscribe( std::move( cb ) );
+
+    if( should_start() )
+        start();
+
+    return ret;
+}
+
+bool options_watcher::should_start() const
+{
+    return ! should_stop();
+}
+
+bool options_watcher::should_stop() const
+{
+    return _on_values_changed.size() == 0 || _options.size() == 0;
+}
+
+void options_watcher::start()
+{
+    if( ! _updater.joinable() ) // If not already started
+    {
+        if( _should_query_for_first_time )
+        {
+            update_options();
+            _should_query_for_first_time = false;
+        }
+        _updater = std::thread( [this]() { thread_loop(); } );
+    }
+}
+
+void options_watcher::stop()
+{
+    if( _updater.joinable() )
+        _updater.join();
+}
+
+void options_watcher::thread_loop()
+{
+    // Checking should_stop because subscriptions can be canceled without us knowing
+    while( ! should_stop() )
+    {
+        std::map< rs2_option, std::shared_ptr< option > > updated_options = update_options();
+
+        // Checking stop conditions after update, if stop requested no need to notify.
+        if( should_stop() )
+            break;
+
+        notify( updated_options );
+
+        // Checking again for stop conditions after callbacks but before sleep.
+        if(  should_stop() )
+            break;
+
+        std::this_thread::sleep_for( _update_interval );
+    }
+}
+
+std::map< rs2_option, std::shared_ptr< option > > options_watcher::update_options()
+{
+    std::map< rs2_option, std::shared_ptr< option > > updated_options;
+
+    std::lock_guard< std::mutex > lock( _mutex );
+
+    for( auto & opt : _options )
+    {
+        try
+        {
+            auto curr_val = opt.second.sptr->query();
+
+            if( opt.second.last_known_value != curr_val )
+            {
+                opt.second.last_known_value = curr_val;
+                updated_options[opt.first] = opt.second.sptr;
+            }
+        }
+        catch( ... )
+        {
+            // Some options cannot be queried all the time (i.e. streaming only)
+        }
+
+        // Checking stop conditions after each query to ensure stop when requested.
+        if( should_stop() )
```

Comment:
This cannot happen:
We locked the mutex.
The destructor, if it is called, cannot do anything and will wait on it.

I think it would be better if we just had an atomic flag `_is_alive` that start() & stop() would directly manipulate.

Reply:
When the mutex is locked options can't be unregistered but subscriptions can be canceled and then we should stop because there is no one to notify of change,

Anyway, added an atomic bool flag

Reply:
Removed atomic bool. Using condition variable to stop sleep when stopping